// typescript-template-language-service-decorator/lib/nodes.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
//
// Original code forked from https://github.com/Quramy/ts-graphql-plugin
Object.defineProperty(exports, "__esModule", { value: true });
var ts = require("typescript/lib/tsserverlibrary");
function relative(from, to) {
    return {
        line: to.line - from.line,
        character: to.line === from.line ? to.character - from.character : to.character,
    };
}
exports.relative = relative;
function findNode(sourceFile, position) {
    function find(node) {
        if (position >= node.getStart() && position < node.getEnd()) {
            return ts.forEachChild(node, find) || node;
        }
    }
    return find(sourceFile);
}
exports.findNode = findNode;
function findAllNodes(sourceFile, cond) {
    var result = [];
    function find(node) {
        if (cond(node)) {
            result.push(node);
            return;
        }
        else {
            ts.forEachChild(node, find);
        }
    }
    find(sourceFile);
    return result;
}
exports.findAllNodes = findAllNodes;
function isTaggedLiteral(node, tags) {
    if (!node || !node.parent) {
        return false;
    }
    if (node.parent.kind !== ts.SyntaxKind.TaggedTemplateExpression) {
        return false;
    }
    var tagNode = node.parent;
    return isTagged(tagNode, tags);
}
exports.isTaggedLiteral = isTaggedLiteral;
function isTagged(node, tags) {
    var text = node.tag.getText();
    return tags.some(function (tag) {
        return text === tag
            || text.startsWith(tag + '.')
            || text.endsWith('.' + tag)
            || text.startsWith(tag + '(');
    });
}
exports.isTagged = isTagged;
