// typescript-template-language-service-decorator/lib/standard-script-source-helper.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
var tsserverlibrary_1 = require("typescript/lib/tsserverlibrary");
var nodes_1 = require("./nodes");
var StandardScriptSourceHelper = /** @class */ (function () {
    function StandardScriptSourceHelper(languageService) {
        this.languageService = languageService;
    }
    StandardScriptSourceHelper.prototype.getNode = function (fileName, position) {
        return nodes_1.findNode(this.languageService.getProgram().getSourceFile(fileName), position);
    };
    StandardScriptSourceHelper.prototype.getAllNodes = function (fileName, cond) {
        var s = this.languageService.getProgram().getSourceFile(fileName);
        return nodes_1.findAllNodes(s, cond);
    };
    StandardScriptSourceHelper.prototype.getLineAndChar = function (fileName, position) {
        var s = this.languageService.getProgram().getSourceFile(fileName);
        return tsserverlibrary_1.getLineAndCharacterOfPosition(s, position);
    };
    StandardScriptSourceHelper.prototype.getOffset = function (fileName, line, character) {
        var s = this.languageService.getProgram().getSourceFile(fileName);
        return tsserverlibrary_1.getPositionOfLineAndCharacter(s, line, character);
    };
    return StandardScriptSourceHelper;
}());
exports.default = StandardScriptSourceHelper;
