// typescript-lit-html-plugin/lib/index.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
//
// Original code forked from https://github.com/Quramy/ts-graphql-plugin
var html_template_language_service_1 = require("./html-template-language-service");
var typescript_template_language_service_decorator_1 = require("typescript-template-language-service-decorator");
var config_1 = require("./config");
var configuration_1 = require("./configuration");
var LanguageServiceLogger = /** @class */ (function () {
    function LanguageServiceLogger(info) {
        this.info = info;
    }
    LanguageServiceLogger.prototype.log = function (msg) {
        this.info.project.projectService.logger.info("[" + config_1.pluginName + "] " + msg);
    };
    return LanguageServiceLogger;
}());
function create(info) {
    var logger = new LanguageServiceLogger(info);
    var config = configuration_1.loadConfiguration(info.config);
    logger.log('config: ' + JSON.stringify(config));
    return typescript_template_language_service_decorator_1.decorateWithTemplateLanguageService(info.languageService, new html_template_language_service_1.default(config, logger), {
        tags: config.tags,
        enableForStringWithSubstitutions: true,
        getSubstitution: function (templateString, start, end) {
            return templateString.slice(start, end);
        },
    }, { logger: logger });
}
module.exports = function (mod) {
    return { create: create };
};
