// typescript-lit-html-plugin/lib/styled-template-language-service.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
//
// Original code forked from https://github.com/Quramy/ts-graphql-plugin
Object.defineProperty(exports, "__esModule", { value: true });
const ts = require("typescript/lib/tsserverlibrary");
const vscode_css_languageservice_1 = require("vscode-css-languageservice");
const vscode = require("vscode-languageserver-types");
const config = require("./config");
const wrapperPre = ':root{\n';
class StyledTemplateLanguageService {
    constructor(configuration) {
        this.configuration = configuration;
    }
    get cssLanguageService() {
        if (!this._cssLanguageService) {
            this._cssLanguageService = vscode_css_languageservice_1.getCSSLanguageService();
            this._cssLanguageService.configure(this.configuration);
        }
        return this._cssLanguageService;
    }
    get scssLanguageService() {
        if (!this._scssLanguageService) {
            this._scssLanguageService = vscode_css_languageservice_1.getSCSSLanguageService();
            this._scssLanguageService.configure(this.configuration);
        }
        return this._scssLanguageService;
    }
    getCompletionsAtPosition(context, position) {
        const doc = this.createVirtualDocument(context);
        const stylesheet = this.cssLanguageService.parseStylesheet(doc);
        const items = this.cssLanguageService.doComplete(doc, this.toVirtualDocPosition(position), stylesheet);
        return translateCompletionItems(items);
    }
    getQuickInfoAtPosition(context, position) {
        const doc = this.createVirtualDocument(context);
        const stylesheet = this.cssLanguageService.parseStylesheet(doc);
        const hover = this.cssLanguageService.doHover(doc, this.toVirtualDocPosition(position), stylesheet);
        if (hover) {
            return this.translateHover(hover, this.toVirtualDocPosition(position), context);
        }
        return undefined;
    }
    getSemanticDiagnostics(context) {
        const doc = this.createVirtualDocument(context);
        const stylesheet = this.scssLanguageService.parseStylesheet(doc);
        return this.translateDiagnostics(this.scssLanguageService.doValidation(doc, stylesheet), doc, context, context.text).filter(x => !!x);
    }
    createVirtualDocument(context) {
        const contents = `${wrapperPre}${context.text}\n}`;
        return {
            uri: 'untitled://embedded.scss',
            languageId: 'scss',
            version: 1,
            getText: () => contents,
            positionAt: (offset) => {
                const pos = context.toPosition(this.fromVirtualDocOffset(offset));
                return this.toVirtualDocPosition(pos);
            },
            offsetAt: (p) => {
                const offset = context.toOffset(this.fromVirtualDocPosition(p));
                return this.toVirtualDocOffset(offset);
            },
            lineCount: contents.split(/n/g).length + 1,
        };
    }
    toVirtualDocPosition(position) {
        return {
            line: position.line + 1,
            character: position.character,
        };
    }
    fromVirtualDocPosition(position) {
        return {
            line: position.line - 1,
            character: position.character,
        };
    }
    toVirtualDocOffset(offset) {
        return offset + wrapperPre.length;
    }
    fromVirtualDocOffset(offset) {
        return offset - wrapperPre.length;
    }
    translateDiagnostics(diagnostics, doc, context, content) {
        const sourceFile = context.node.getSourceFile();
        return diagnostics.map(diag => this.translateDiagnostic(diag, sourceFile, doc, context, content));
    }
    translateDiagnostic(diagnostic, file, doc, context, content) {
        // Make sure returned error is within the real document
        if (diagnostic.range.start.line === 0
            || diagnostic.range.start.line >= doc.lineCount
            || diagnostic.range.start.character >= content.length) {
            return undefined;
        }
        const start = context.toOffset(this.fromVirtualDocPosition(diagnostic.range.start));
        const length = context.toOffset(this.fromVirtualDocPosition(diagnostic.range.end)) - start;
        const code = typeof diagnostic.code === 'number' ? diagnostic.code : 9999;
        return {
            code,
            messageText: diagnostic.message,
            category: translateSeverity(diagnostic.severity),
            file,
            start,
            length,
            source: config.pluginName,
        };
    }
    translateHover(hover, position, context) {
        const contents = [];
        const convertPart = (hoverContents) => {
            if (typeof hoverContents === 'string') {
                contents.push({ kind: 'unknown', text: hoverContents });
            }
            else if (Array.isArray(hoverContents)) {
                hoverContents.forEach(convertPart);
            }
            else {
                contents.push({ kind: 'unknown', text: hoverContents.value });
            }
        };
        convertPart(hover.contents);
        const start = context.toOffset(this.fromVirtualDocPosition(hover.range ? hover.range.start : position));
        return {
            kind: ts.ScriptElementKind.unknown,
            kindModifiers: '',
            textSpan: {
                start,
                length: hover.range ? context.toOffset(this.fromVirtualDocPosition(hover.range.end)) - start : 1,
            },
            displayParts: [],
            documentation: contents,
            tags: [],
        };
    }
}
exports.default = StyledTemplateLanguageService;
function translateCompletionItems(items) {
    return {
        isGlobalCompletion: false,
        isMemberCompletion: false,
        isNewIdentifierLocation: false,
        entries: items.items.map(translateCompetionEntry),
    };
}
function translateCompetionEntry(item) {
    return {
        name: item.label,
        kindModifiers: 'declare',
        kind: item.kind ? translateionCompletionItemKind(item.kind) : ts.ScriptElementKind.unknown,
        sortText: '0',
    };
}
function translateionCompletionItemKind(kind) {
    switch (kind) {
        case vscode.CompletionItemKind.Method:
            return ts.ScriptElementKind.memberFunctionElement;
        case vscode.CompletionItemKind.Function:
            return ts.ScriptElementKind.functionElement;
        case vscode.CompletionItemKind.Constructor:
            return ts.ScriptElementKind.constructorImplementationElement;
        case vscode.CompletionItemKind.Field:
        case vscode.CompletionItemKind.Variable:
            return ts.ScriptElementKind.variableElement;
        case vscode.CompletionItemKind.Class:
            return ts.ScriptElementKind.classElement;
        case vscode.CompletionItemKind.Interface:
            return ts.ScriptElementKind.interfaceElement;
        case vscode.CompletionItemKind.Module:
            return ts.ScriptElementKind.moduleElement;
        case vscode.CompletionItemKind.Property:
            return ts.ScriptElementKind.memberVariableElement;
        case vscode.CompletionItemKind.Unit:
        case vscode.CompletionItemKind.Value:
            return ts.ScriptElementKind.constElement;
        case vscode.CompletionItemKind.Enum:
            return ts.ScriptElementKind.enumElement;
        case vscode.CompletionItemKind.Keyword:
            return ts.ScriptElementKind.keyword;
        case vscode.CompletionItemKind.Color:
            return ts.ScriptElementKind.constElement;
        case vscode.CompletionItemKind.Reference:
            return ts.ScriptElementKind.alias;
        case vscode.CompletionItemKind.File:
            return ts.ScriptElementKind.moduleElement;
        case vscode.CompletionItemKind.Snippet:
        case vscode.CompletionItemKind.Text:
        default:
            return ts.ScriptElementKind.unknown;
    }
}
function translateSeverity(severity) {
    switch (severity) {
        case vscode.DiagnosticSeverity.Information:
        case vscode.DiagnosticSeverity.Hint:
            return ts.DiagnosticCategory.Message;
        case vscode.DiagnosticSeverity.Warning:
            return ts.DiagnosticCategory.Warning;
        case vscode.DiagnosticSeverity.Error:
        default:
            return ts.DiagnosticCategory.Error;
    }
}
